### PR TITLE
Make CaseBlock's default date_opened a datetime not a date

### PR DESCRIPTION
**Pull request by dannyroberts**
https://dimagi-dev.atlassian.net/browse/SAAS-10795

##### SUMMARY
This fixes the `date_opened` field on cases that are opened using the `CaseBlock` utility. Previously they were just a date, which was arbitrarily interpreted as meaning midnight UTC on that date. Now it will be a datetime like normal cases created from web apps or a phone.

##### FEATURE FLAG
None

##### RISK ASSESSMENT / QA PLAN
Other things do use this, but I don't think anything relies on this (strictly inferior) behavior (other than tests)

### CODE REVIEW

**Review by ctsims — APPROVED** [review]

**Comment by dannyroberts** [comment]
I think this change requires some explanation given the wide range of files that it touches now.

In the first iteration of this change, I wanted to change `CaseBlock` to just not insert `<date_opened>YYYY-MM-DD</date_opened>` in the update block of every case block with a create block. It's redundant with the default except it leaves our opened_on field with a date + `00:00:00.000000Z` instead of a full-precision `datetime`.

However, just making that ~1 line change of course broke dozens if not hundreds of tests. So instead, I'm keeping an option to use the old behavior, changing _all calls to `CaseBlock`_ to explicitly use the old behavior, and then changing just the case importer calls to CaseBlock back to the new default. A much smaller change would be to make the deprecated behavior the default and have the case importer calls opt in to use it, but since using the deprecated behavior really does/can result in unintuitive loss of precision, I wanted to leave the code in a state that strongly suggests these behaviors be updated not to use the deprecated behavior. Going through and validating that this was safe to use everywhere and then fixing the hundreds of tests felt error prone and out of scope for this change. And so what I did in this PR was the smallest change that I could think of that still puts us on the right longer-term path.

Open to feedback

File: corehq/ex-submodules/casexml/apps/case/mock/case_block.py [hunk 1]
```diff
@@ -35,8 +35,7 @@ def __init__(self, case_id, date_modified=None, user_id=undefined,
         self.update = copy.copy(update) if update else {}
         now = datetime.utcnow()
         self.date_modified = date_modified or now
-        self.date_opened = (now if create and date_opened is CaseBlock.undefined
-                            else date_opened)
+        self.date_opened = date_opened
```
Comment (mkangia) [inline corehq/ex-submodules/casexml/apps/case/mock/case_block.py]: "date_opened automatically is set to date_modified"
Hey @dannyroberts , where does this happen?

**Comment by proteusvacuum** [comment]
This brought back many memories (mostly because the words `case` and `date_opened` generate fear in my heart): https://docs.google.com/presentation/d/1hhs4Hi-ms7LaToOMINVzyAjF_N8LEalP4qlop42oqBA/edit#slide=id.p 

It sounds like you aren't actually changing the way things are getting synced to the phone, and you are saying that we are actually storing `date_opened` as a datetime normally anyway, so this seems fine (and I checked - since that retro we do things "right" now). 

Thought I'd bring back that little part of history anyway #throwbackMonday.

**Review by millerdev — APPROVED** [review]
I agree that making the default/easiest use of `CaseBlock` do the sensible thing and changing all old uses to do something obviously deprecated is a good approach.

File: corehq/ex-submodules/casexml/apps/case/mock/case_block.py [hunk 2]
```diff
@@ -25,7 +25,11 @@ class CaseBlock(object):
     def __init__(self, case_id, date_modified=None, user_id=undefined,
                  owner_id=undefined, external_id=undefined, case_type=undefined,
                  case_name=undefined, create=False, date_opened=undefined, update=None,
-                 close=False, index=None, strict=True):
+                 close=False, index=None, strict=True, date_opened_deprecated_behavior=False):
+        """
+        When `date_opened_deprecated_behavior`, a date_opened YYYY-MM-DD value is inserted on new cases.
+        This is deprecated behavior, because it prevents the superior default behavior from kicking in.
```
Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/case/mock/case_block.py:31]: where does this other default behaviour happen?

Comment (dannyroberts) [reply corehq/ex-submodules/casexml/apps/case/mock/case_block.py:31]: We do it explicitly during case processing. In the sql backend for example I think it'd done here: https://github.com/dimagi/commcare-hq/blob/f7b136fb44787e5fdacde7dce5c922880e9a01c8/corehq/form_processor/backends/sql/update_strategy.py#L138-L139

File: corehq/ex-submodules/casexml/apps/case/mock/case_block.py (hunk 1 above)
Comment (dannyroberts) [reply corehq/ex-submodules/casexml/apps/case/mock/case_block.py]: Hey sorry missed this yesterday after the diff became outdated. Answered it later here when Simon asked it again: https://github.com/dimagi/commcare-hq/pull/27875#discussion_r444253403

**Comment by dannyroberts** [comment]
OK finally ✅. Tests pass, reviewers approved, local testing works, the theory of what I'm doing checks out, and the changes are very rote despite the big footprint. Merging!

File: corehq/ex-submodules/casexml/apps/case/mock/case_block.py (hunk 1 above)
Comment (mkangia) [reply corehq/ex-submodules/casexml/apps/case/mock/case_block.py]: Cool, thanks @dannyroberts

**Pull request merged**